include/ginkgo/core/log/profiler_hook.hpp

```diff
@@ -0,0 +1,246 @@
+                                           const LinOp* x) const override;
+
+    /* LinOpFactory events */
+    void on_linop_factory_generate_started(const LinOpFactory* factory,
+                                           const LinOp* input) const override;
+
+    void on_linop_factory_generate_completed(
+        const LinOpFactory* factory, const LinOp* input,
+        const LinOp* output) const override;
+
+    /* Criterion events */
+    void on_criterion_check_started(const stop::Criterion* criterion,
+                                    const size_type& num_iterations,
+                                    const LinOp* residual,
+                                    const LinOp* residual_norm,
+                                    const LinOp* solution,
+                                    const uint8& stopping_id,
+                                    const bool& set_finalized) const override;
+
+    void on_criterion_check_completed(
+        const stop::Criterion* criterion, const size_type& num_iterations,
+        const LinOp* residual, const LinOp* residual_norm,
+        const LinOp* solution, const uint8& stopping_id,
+        const bool& set_finalized, const array<stopping_status>* status,
+        const bool& one_changed, const bool& all_stopped) const override;
+
+    void on_criterion_check_completed(
+        const stop::Criterion* criterion, const size_type& num_iterations,
+        const LinOp* residual, const LinOp* residual_norm,
+        const LinOp* implicit_sq_resnorm, const LinOp* solution,
+        const uint8& stopping_id, const bool& set_finalized,
+        const array<stopping_status>* status, const bool& one_changed,
+        const bool& all_stopped) const override;
+
+    bool needs_propagation() const override;
+
+    /**
+     * Sets the name for an object to be profiled. Every instance of that object
+     * in the profile will be replaced by the name instead of its runtime type.
+     * @param obj  the object
+     * @param name  its name
+     */
+    void set_object_name(const PolymorphicObject* obj, std::string name);
+
+    /**
+     * Should the events call executor->synchronize on operations and
+     * copy/allocation? This leads to a certain overhead, but makes the
+     * execution timeline of kernels synchronous.
+     */
+    void set_synchronization(bool synchronize);
+
+    /** The Ginkgo yellow background color as packed 32 bit ARGB value. */
+    constexpr static uint32 color_yellow_argb = 0xFFFFCB05U;
+
+    /**
+     * Creates a logger annotating Ginkgo events with TAU ranges via PerfStubs.
+     * Since TAU requires global initialization and finalization, this will only
+     * create a single global instance that is destroyed at program exit.
+     * @param initialize  Should we call TAU's initialization and finalization
+     *                    functions, or does the application take care of it?
+     */
+    static std::shared_ptr<ProfilerHook> create_tau(bool initialize = true);
```

Comment:
that requires people to use gko::share, which is a bit annoying if you want to add names to objects, since loggers can only be used via shared pointers

Reply:
That is true, but that holds also for all of our other loggers, especially `Convergence`. I just would like to have a single consistent way of using our loggers. But I won't hold up the PR over this.

Reply:
done